### PR TITLE
[POAE7-2703] add date/time extract func in nextgen

### PR DESCRIPTION
**Pull request by kd97**
### What changes were proposed in this pull request?

extract date/time  support in nextgen
### Why are the changes needed?

nextgen feature

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

UTs

### CODE REVIEW

File: src/cider/function/scalar/RuntimeFunctions.cpp [hunk 1]
```diff
@@ -1480,7 +1480,7 @@ extern "C" ALWAYS_INLINE int32_t extract_str_len_arrow(int8_t* offset_buffer,
 }
 
 #include "exec/nextgen/context/ContextRuntimeFunctions.h"
+#include "exec/nextgen/function/CiderDateFunctions.cpp"
```
Comment (spevenhe) [inline src/cider/function/scalar/RuntimeFunctions.cpp:1483]: why these files are included at end of the file?

File: src/cider/exec/nextgen/function/CiderDateFunctions.cpp [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+  // Return number of seconds since 1 January 1970.
+  int64_t getTime() const { return getDate() * kSecsPerDay + offset_unit_.sod; }
+
+  // Return dates since 1 January 1970.
+  int32_t getDate() const {
+    return (kEpochAdjustedDays + era_ * kDaysPer400Years + offset_unit_.doe);
+  }
+
+  int64_t extractYear() const {
+    return 2000 + era_ * 400 + offset_unit_.yoe + (MARJAN <= offset_unit_.doy);
```
Comment (spevenhe) [inline src/cider/exec/nextgen/function/CiderDateFunctions.cpp:113]: only support years after 2000?

File: src/cider/exec/nextgen/function/CiderDateFunctions.cpp [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+ */
+
+#include "function/datetime/DateAdd.h"
+
+namespace {
+
+struct OffsetUnit {
+  unsigned dom;  // day-of-month (0-based)
+  unsigned sod;  // second-of-day
+  unsigned doe;  // day-of-era
```
Comment (spevenhe) [inline src/cider/exec/nextgen/function/CiderDateFunctions.cpp:29]: what does era mean?

File: src/cider/function/scalar/RuntimeFunctions.cpp (hunk 1 above)
Comment (yma11) [reply src/cider/function/scalar/RuntimeFunctions.cpp:1483]: Seems new added function files for nextgen are put here.

**Comment by yma11** [comment]
LGTM. But need fix lint issue first.

File: src/cider/function/scalar/RuntimeFunctions.cpp (hunk 1 above)
Comment (kd97) [reply src/cider/function/scalar/RuntimeFunctions.cpp:1483]: I think this is just to distinguish which header files (beginning of the file) actually need to be included by `RuntimeFunctions.cpp`, which (end of the file) are new added nextgen function modules. maybe can refactor this file and find better way(like a placeholder header file above `RuntimeFunctions.cpp`) to consolidate these files after fully migrated to nextgen.

File: src/cider/exec/nextgen/function/CiderDateFunctions.cpp (hunk 3 above)
Comment (kd97) [reply src/cider/exec/nextgen/function/CiderDateFunctions.cpp:29]: kEpochAdjustedDays is days from 01.01.1970 to 01.03.2000. so era is a unit divided by 400 years, and the 0th era starts from 01.03.2000.

https://github.com/intel/BDTK/blob/1a167f0867cc5d9db0d66aefa4f4d545a79eb614/src/cider/exec/nextgen/function/CiderDateFunctions.cpp#L73-L76

File: src/cider/exec/nextgen/function/CiderDateFunctions.cpp (hunk 2 above)
Comment (kd97) [reply src/cider/exec/nextgen/function/CiderDateFunctions.cpp:113]: era_ will be a negative value if date is before 2000.

**Review by jikunshang — DISMISSED** [review]

**Comment by kd97** [comment]
rebase code, pls help review @jikunshang @haiweizh

**Review by jikunshang — DISMISSED** [review]

**Review by spevenhe — DISMISSED** [review]

**Review by haiweizh — DISMISSED** [review]

File: src/cider/function/scalar/RuntimeFunctions.cpp [hunk 4]
```diff
@@ -1480,8 +1480,8 @@ extern "C" ALWAYS_INLINE int32_t extract_str_len_arrow(int8_t* offset_buffer,
...
 #include "exec/nextgen/function/CiderSetFunctions.cpp"
 #include "exec/nextgen/function/CiderStringFunctions.cpp"
 #include "exec/nextgen/operators/OperatorRuntimeFunctions.h"
-#include "function/aggregate/CiderRuntimeFunctions.h"
-#include "function/datetime/CiderDateFunctions.cpp"
+#include "function/aggregate/CiderRuntimeFunctions.h"
```
Comment (yma11) [inline src/cider/function/scalar/RuntimeFunctions.cpp:1487]: add empty line?

**Review by yma11 — DISMISSED** [review]

**Review by jikunshang — APPROVED** [review]

**Review by spevenhe — APPROVED** [review]

**Review by jikunshang — APPROVED** [review]

**Pull request merged**